File: go/email/email.go
// Package email validates, normalizes, and formats e-mail addresses.
package email

import (
	"regexp"
	"strings"
)

// Sourced from the WHATWG HTML Living Standard's email state regex (used by
// browsers to validate <input type="email">). All quantifiers are bounded,
// so it cannot suffer catastrophic backtracking.
var emailRegex = regexp.MustCompile(
	`^[a-zA-Z0-9.!#$%&'*+/=?^_` + "`" + `{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$`,
)

const maxEmailLength = 254
const maxLocalPartLength = 64

// Normalize trims whitespace and lowercases value.
func Normalize(value string) string {
	return strings.ToLower(strings.TrimSpace(value))
}

// IsValid reports whether value is a structurally valid e-mail address,
// per the WHATWG regular expression plus RFC 5321 length limits.
func IsValid(value string) bool {
	email := Normalize(value)

	if len(email) == 0 || len(email) > maxEmailLength {
		return false
	}

	localPart := email
	if i := strings.IndexByte(email, '@'); i >= 0 {
		localPart = email[:i]
	}

	if localPart == "" || len(localPart) > maxLocalPartLength {
		return false
	}

	return emailRegex.MatchString(email)
}

// Format returns the same canonical value as Normalize. Unlike CPF/CNPJ/
// CEP/Phone, an e-mail address has no visual mask to apply.
func Format(value string) string {
	return Normalize(value)
}

File: src/email/email.ts
// Sourced from the WHATWG HTML Living Standard's email state regex
// (used by browsers to validate <input type="email">). All quantifiers are
// bounded, so it cannot suffer catastrophic backtracking.
const EMAIL_REGEX =
  /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$/;

const MAX_EMAIL_LENGTH = 254;
const MAX_LOCAL_PART_LENGTH = 64;

function normalize(value: string): string {
  return value.trim().toLowerCase();
}

function isValid(value: string): boolean {
  const email = normalize(value);

  if (email.length === 0 || email.length > MAX_EMAIL_LENGTH) {
    return false;
  }

  const [localPart] = email.split("@");

  if (!localPart || localPart.length > MAX_LOCAL_PART_LENGTH) {
    return false;
  }

  return EMAIL_REGEX.test(email);
}

// An email address has no punctuation mask to apply (unlike CPF/CNPJ/CEP/
// Phone), so the formatted form is the same canonical value as normalize().
function format(value: string): string {
  return normalize(value);
}

export const Email = {
  isValid,
  normalize,
  format,
};
